src/core/dag.cpp

```diff
@@ -0,0 +1,1366 @@
+#include "quartz/dag/dag.h"
+#include "quartz/gate/gate.h"
+#include "quartz/context/context.h"
+
+#include <algorithm>
+#include <cassert>
+#include <charconv>
+#include <unordered_set>
+#include <utility>
+
+using namespace quartz;
```

Comment:
I actually like the design of also putting the implementation into `namespace quartz { ... }` better.

Of course `using namespace quartz;` will also work and we can merge this PR as is and discuss the design later.

Reply:
Yes, I think so, I'll modify them later.